hoodie-client/src/main/java/com/uber/hoodie/func/EmbeddedTimelineServer.java

```diff
@@ -0,0 +1,136 @@
+    Pair<String, Integer> hostPort = NetworkUtils.getAvailableServerPortWithHostname();
+    this.hostAddr = hostPort.getKey();
+    this.serverPort = hostPort.getRight();
+    this.config = config;
+    this.hadoopConf = new SerializableConfiguration(hadoopConf);
+    this.viewManager = createViewManager();
+  }
+
+  public <T extends HoodieRecordPayload> boolean preloadNeededPartitions(JavaSparkContext jsc,
+      List<String> partitions) {
+    // Pre populate File System View of the table
+    logger.info("Pre-populating file system view in driver. Total Partitions :" + partitions.size());
+    try {
+      new ForkJoinPool(config.getMaxFileStatusListerThreads()).submit(() ->
+          partitions.parallelStream().forEach(
+              partition -> {
+                // logger.info("Finding latest slices in partition " + partition);
+                long numFileSlices =
+                    viewManager.getFileSystemView(config.getBasePath()).getLatestFileSlices(partition).count();
+                logger.info("Found " + numFileSlices + " latest slices in partition " + partition);
+              }))
+          .get();
+    } catch (InterruptedException | ExecutionException e) {
+      logger.error("Got exception when pre populating file system view", e);
+      return false;
+    }
+    logger.info("Pre-population of File system view finished for partitions :" + partitions.size());
+    setHostAddrFromSparkConf(jsc.getConf());
+    return true;
+  }
+
+  private FileSystemViewManager createViewManager() {
+    FileSystemViewStorageConfig sConf = buildStorageConfigForServer();
+    return FileSystemViewManager.createViewManager(hadoopConf, sConf);
+  }
+
+  private FileSystemViewStorageConfig buildStorageConfigForServer() {
+    // Using passed-in configs to build view storage configs
+    FileSystemViewStorageConfig.Builder builder =
+        FileSystemViewStorageConfig.newBuilder().fromProperties(
+            config.getClientSpecifiedViewStorageConfig().getProps());
+    FileSystemViewStorageType storageType = builder.build().getStorageType();
+    if (storageType.equals(FileSystemViewStorageType.REMOTE_ONLY)
+        || storageType.equals(FileSystemViewStorageType.REMOTE_FIRST)) {
+      // Reset to default if set to Remote
+      builder.withStorageType(FileSystemViewStorageType.MEMORY);
+    }
+    return builder.build();
+  }
+
+  public void startServer() throws IOException {
+    logger.info("Starting embedded timeline server on port " + serverPort);
+    server = new TimelineServer(serverPort, viewManager);
+    server.startService();
+  }
+
+  private void setHostAddrFromSparkConf(SparkConf sparkConf) {
+    String hostAddr = sparkConf.get("spark.driver.host", null);
+    if (hostAddr != null) {
+      logger.info("Overriding hostIp to (" + hostAddr + ") found in spark-conf. It was " + this.hostAddr);
+      this.hostAddr = hostAddr;
+    } else {
+      logger.warn("Unable to find driver bind address from spark config");
```

Comment:
Should this throw an exception ? What is the expected behavior if this happens ?